internal/pkg/instrumentation/bpf/go.opentelemetry.io/auto/sdk/bpf/probe.bpf.c

```diff
@@ -0,0 +1,196 @@
+
+	void *parent_span_context = get_argument(ctx, 5);
+	long rc = write_span_context(parent_span_context, &otel_span.psc);
+	if (rc != 0) {
+		bpf_printk("failed to write parent span context: %ld", rc);
+	}
+
+	if (!is_sampled(params.sc)) {
+		// Default SDK behaviour is to sample everything. Only set the sampled
+		// value to false when needed.
+		void *sampled_ptr_val = get_argument(ctx, 6);
+		if (sampled_ptr_val == NULL) {
+			bpf_printk("nil sampled pointer");
+		} else {
+			bool false_val = false;
+			rc = bpf_probe_write_user(sampled_ptr_val, &false_val, sizeof(bool));
+			if (rc != 0) {
+				bpf_printk("bpf_probe_write_user: failed to write sampled value: %ld", rc);
+			} else {
+				bpf_printk("wrote sampled value");
+			}
+		}
+	}
+
+	void *span_context_ptr_val = get_argument(ctx, 7);
+	rc = write_span_context(span_context_ptr_val, &otel_span.sc);
+	if (rc != 0) {
+		bpf_printk("failed to write span context: %ld", rc);
+	}
+
+	void *span_ptr_val = get_argument(ctx, 4);
+    bpf_map_update_elem(&active_spans_by_span_ptr, &span_ptr_val, &otel_span, 0);
+    start_tracking_span(go_context.data, &otel_span.sc);
+
+	return 0;
+}
+
+// This instrumentation attaches a uprobe to the following function:
+// func (*span) ended(buf []byte) {}
+// https://github.com/open-telemetry/opentelemetry-go-instrumentation/blob/effdec9ac23e56e9e9655663d386600e62b10871/sdk/trace.go#L133-L136
+SEC("uprobe/Span_ended")
+int uprobe_Span_ended(struct pt_regs *ctx) {
+    void *span_ptr = get_argument(ctx, 1);
+    struct otel_span_t *span = bpf_map_lookup_elem(&active_spans_by_span_ptr, &span_ptr);
+    if (span == NULL) {
+        return 0;
+    }
+    stop_tracking_span(&span->sc, &span->psc);
+    bpf_map_delete_elem(&active_spans_by_span_ptr, &span_ptr);
+
+	// Do not output un-sampled span data.
+	if (!is_sampled(&span->sc)) return 0;
+
+	u64 len = (u64)get_argument(ctx, 3);
+	if (len > MAX_SIZE) {
+		bpf_printk("span data too large: %d", len);
+		return -1;
+	}
+	if (len == 0) {
+		bpf_printk("empty span data");
+		return 0;
+	}
+
+	void *data_ptr = get_argument(ctx, 2);
+	if (data_ptr == NULL) {
+		bpf_printk("empty span data");
+		return 0;
+	}
+
+	u32 key = 0;
+	struct event_t *event = bpf_map_lookup_elem(&new_event, &key);
+	if (event == NULL) {
+		bpf_printk("failed to initialize new event");
+		return -2;
+	}
+	event->size = (u32)len;
+
+	if (event->size < MAX_SIZE) {
+		long rc = bpf_probe_read(&event->data, event->size, data_ptr);
+		if (rc < 0) {
+			bpf_printk("failed to read encoded span data");
+			return -3;
+		}
+	} else {
+		bpf_printk("read too large: %d", event->size);
+		return -4;
+	}
+
+	// Do not send the whole size.buf if it is not needed.
+	u64 size = sizeof(event->size) + event->size;
+	// Make the verifier happy, ensure no unbounded memory access.
+	if (size < sizeof(struct event_t)+1) {
+		return bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU, event, size);
+	}
+	bpf_printk("write too large: %d", event->size);
```

Comment:
Maybe this will make the verifier happy with the `output_span_event` function

```suggestion
         if (event->size > MAX_SIZE) {
             bpf_printk("write too large: %d", event->size);
             return -5;
         }
         // assuming MAX_SIZE is a power of 2
         u64 payload_size = size & (MAX_SIZE - 1)
         output_span_event(ctx, event, sizeof(event->size) + payload_size,  &span->sc)
```

Reply:
I think a version of this could work. But, I'm wondering if we need to use `output_span_event`. We are checking the sampling decision at the top of the function and avoiding unnecessary work there. Given that `output_span_event` just wraps that check and `bpf_perf_event_output` it seems like it will just introduce redundant steps.

Reply:
Right, I missed the check for the is_sampled